src/Compatibility/Microsoft.DotNet.ApiCompatibility/ValidationSuppressions/SuppressionEngine.cs

```diff
@@ -0,0 +1,208 @@
+
+#nullable enable
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Xml.Serialization;
+
+namespace Microsoft.DotNet.ValidationSuppression
+{
+    /// <summary>
+    /// Collection of Suppressions which is able to add suppressions, check if a specific error is suppressed, and write all suppressions
+    /// down to a file. The engine is thread-safe.
+    /// </summary>
+    public class SuppressionEngine
+    {
+        protected HashSet<Suppression> _validationSuppressions;
+        private ReaderWriterLockSlim _readerWriterLock = new();
+
+        protected SuppressionEngine(string validationSuppressionFile)
+        {
+            _validationSuppressions = ParseValidationSuppressionBaseline(validationSuppressionFile);
+        }
+
+        protected SuppressionEngine()
+        {
+            _validationSuppressions = new HashSet<Suppression>(new SuppressionComparer());
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="diagnosticId">The diagnostic ID of the error to check.</param>
+        /// <param name="target">The target of where the <paramref name="diagnosticId"/> should be applied.</param>
+        /// <param name="left">Optional. The left operand in a APICompat error.</param>
+        /// <param name="right">Optional. The right operand in a APICompat error.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(string? diagnosticId, string? target, string? left = null, string? right = null)
+        {
+            var suppressionToCheck = new Suppression()
+            {
+                DiagnosticId = diagnosticId,
+                Target = target,
+                Left = left,
+                Right = right
+            };
+            return IsErrorSuppressed(suppressionToCheck);
+        }
+
+        /// <summary>
+        /// Checks if the passed in error is suppressed or not.
+        /// </summary>
+        /// <param name="error">The <see cref="Suppression"/> error to check.</param>
+        /// <returns><see langword="true"/> if the error is already suppressed. <see langword="false"/> otherwise.</returns>
+        public bool IsErrorSuppressed(Suppression error)
+        {
+            _readerWriterLock.EnterReadLock();
+            try
+            {
+                if(_validationSuppressions.Contains(error))
+                {
+                    return true;
+                }
+                else
+                {
+                    // See if the error is globally suppressed by checking if the same diagnosticid and target are entered
+                    // without any left and right.
+                    if ((error.DiagnosticId == null || !error.DiagnosticId.StartsWith("pkv", StringComparison.InvariantCultureIgnoreCase)) 
```

Comment:
can you explain this condition ```((error.DiagnosticId == null || !error.DiagnosticId.StartsWith("pkv", StringComparison.InvariantCultureIgnoreCase))```

Reply:
Sure, the idea here is that imagine you have the case of an API Compat validation error that applies to multiple left and multiple right, for example: a member that used to exist in your previous package in netstandard2.0 got removed in your new package for both netstandard2.0 and net6.0 assets. This means you will get 2 APICompat validation errors, one for each pair. What this condition supports, is that if you decide  in your suppression file to add a suppression that only has a diagnostic id and a target (in this case the member that was removed) then it will be treated as a global suppression and will be suppressed for both of the cases.

Reply:
You can check out this unit test as an example, which is exercising this if check:
https://github.com/dotnet/sdk/pull/18185/files/107da235e17bce8a0fcd18460c15f18782b703ee#diff-df90c3b499c9c47bff81aaffef3af26830f68ae2b57ae6efb4a9059064021ecdR74-R81

Reply:
Finally, this only relates to ApiCompat errors (since they are the only ones that compare left's and right's) which is why I have the second part of the condition which checks that the diagnosticId is not a PKV (package validation error)

Reply:
Wouldn't this never be hit? I thought that `Equality` of the suppressions encounters for null/empty `Left` and `Right`? So the previous if would be true?

Reply:
Ah, I see with the test example. Never mind, cool that we support this "global" suppression.